proposals/helm-repo-container-registry-convergence/002-chart-registry-repo.md

```diff
@@ -0,0 +1,23 @@
+## Helm Charts Align With Registry Repositories
+Container images are referneced with a multi-part URL:
+
+- [**login url**] / [**namespace**] / [**repo**] : [**tag**]
+- [**namespace**] / [**repo**] : [**tag** ] are equally referred to as an "image"
+
+Different clouds implement the unique login url in a few different forms, while effectively meaning the same thing:
+
+| Cloud | Format | eg: image |
+|-----|-----|-----|
+| Docker Hub | [**org**\|**user** ] / [**image**] : [**tag**] **\*1**| [contoso/marketing-fall2018-web:aab1]() |
+| Azure Container Registry (acr) | [**uniqueId**].azurecr.io / [**namespace**] / [**image**] : [**tag**] | contoso.azurecr.io/marketing/fall2018/web:aab1 |
+| Google Container Registry (gcr) | gcr.io/[**uniqueId**] / [**image**] : [**tag**] | gcr.io/marketing/fall2018-web:aab1 |
+| AWS Container Registry (ecr) | [**aws_account_id**].dkr.ecr.region.amazonaws.com/ [**image**] : [**tag**] | 123abc456def.dkr.ecr.region.amazonaws.com/marketing/fall2018-web:aab1 |
+
+1. Docker Hub has a default login url [docker.io]()
+1. Docker Hub [**org**\|**user** ] is a [**namespace**] which is currently limted to two nodes.  
+  *The [Microsoft Container Registry (mcr)](mcr.microsoft.com) has implemented a syndicated catalog representing a product hierarchy within the docker store, and looks for Docker to expand this for other partners with deep image catalogs.*
+
+## Helm References
+
+Proposal aligns Helm Charts with the same referencing scheme:
+- [**login url**] / [**namespace**] / [**chart**] : [**version/tag**]
```

Comment:
Channels are a very interesting concept, specifically as chart owners wish to flight updates to a specific version, and they provide a preview for what's coming. 
The value I see for mutable tags(versions) is the ability to get security updates to a specific version without direct intervention:
- I deploy version 1 of a thing
- Version 2 of the thing is made available. By calling it version 2, it's presumed to be breaking changes, and I (as the developer/ops person) should opt into version 2
- A security flaw is found in sub-particle-a that both Version 1 and 2 reference in their charts.
- The security fix is flighted to the preview channel for both version 1 & 2
- After approval, thing: version 1 and thing:2 are updated. 
- My company is watching changes to version 1, and quickly takes the update, tests it, and deploys the updated "stable" version.

In the above example, the thing owners were able to flight a security fix to version 1 & 2. The users of thing:1 were able to get updates, without having to involve development teams to make specific changes. 

The great thing about additional channels is an org can also run automated tests on the preview channel. If any errors occur, they can automate reporting of issues. 

If i reference a sha directly, I'm always bound to that very specific version, with no ability to move forward for automated patching. I wrote about this here: https://stevelasker.blog/2018/03/01/docker-tagging-best-practices-for-tagging-and-versioning-docker-images/ 

I think channels and stable tags are better together, but not mutual exclusive.

Reply:
> The value I see for mutable tags(versions) is the ability to get security updates to a specific version without direct intervention:

That exactly the purpose of the channel, and was automatic upgrade was the main focus of CoreOS, this why we created channels for the helm-charts repositories.
 You can of course use channels like version/tags, or 'allow' mutable tags (like in git, you protect/unprotect force-push for branches/tags) but default/recommended usage would be:

channel: mutable
tag: non-mutable

Automatic upgrades should be opt-in, not a side effect.
- I deploy **channel** `1-stable` of a thing (points to the version/tag `1.2.0`)
- Version 2 of the thing is made available. By calling it version/tag `2.0.0`, it's presumed to be breaking changes, and I (as the developer/ops person) should opt into channel `2-stable`
- A security flaw is found in sub-particle-a that both Version `1.2.0` and `2.0.0` reference in their charts.
The security fix is flighted to both channels `1-beta` & `2-beta`,
- After approval, version/tags `1.2.1` and  `2.0.1` are pushed,  channels `1-stable` is updated to points to `1.2.1`, and channel `2-beta` is updated to `2.0.1`

- My company is watching changes to version `1-stable`, and quickly takes the update, tests it, and deploys the updated "stable" version.

In term of use experience 
The chart maintainer it is doing:

1. Release the security fix in the beta channels

``` shell
# Push 1.2.1 to the beta channel
$ helm push quay.io/myorg  --channel 1-beta
>>> Release '1.2.1' added to the '1-beta' channel

# Push 2.0.1 to the beta channel
$ helm push quay.io/myorg  --channel 2-beta
>>> Release '2.0.1' added to the '2-beta' channel
```
2. Review channels:
```
$ helm registry channel quay.io/myorg/thing
channel    release
---------  ---------
1-stable     1.2.0
1-beta        1.2.1
2-stable      2.0.0
2-beta        2.0.1
```
2. Once confirmed / approved, promote from `beta` to  `stable`
``` shell
$ helm push quay.io/myorg  --channel 1-stable
>>> Release '1.2.1' added to the '1-stable' channel
$ helm push quay.io/myorg  --channel 2-stable
>>> Release '2.0.1' added to the '2-stable' channel
```

For the application use the experience is either:
- manually pull versions
```
helm registry install quay.io/org/thing@v1.2.1
```

- Trust the channels and get automatic latest versions
```
helm registry install quay.io/org/thing:1-stable
```
But in the end the user has the choice, and he can trust that @1.2.1 will always be the exact same version.

Reply:
I see where your going. It's basically a more explicit way to opt into updates for a stable tag. 
The thing that strikes me with this is:
- customers are already struggling with stable and unique tags
- they don't seem to be understanding the wave of problems ahead of them for patching containers
- when I tell them about it, they either think containers aren't as great as they hoped (looking backwards), or they get it and love the ability to patch, test, then deploy a container
- channels adds yet another concept they are struggling to keep up with
- how can we get this into distribution across all registries so it's a common thing. Unless it's common, I struggle with it getting great adoption

I do like the expectations. Just struggle with getting it to common usage.

Is this something we can layer in, as customers feel the pain?

Reply:
> channels adds yet another concept they are struggling to keep up with

We can completely layer that, and not even talk about channels to begin with. 
Channels are not a key feature, nor required to use a registry, then it should not be a big deal if vendor don't adopt it at the same time.

I'm bringing that early, because it could change the cli UX.
With app-registry, after several iteration we landed with:
```
helm pull domain/org/app:channel
helm pull domain/org/app@tag
helm pull domain/org/app@sha256
```

If we delay channels for let's say12 months, and everyone is using the syntax:
```
helm pull domain/org/app:tag
```
it will be impossible to switch to `:channel` / `@tag`

So around channel, if the concept make sense to you, the only question to answer early is: 

- how users will pull or deploy from a channel via the CLI ?

ps: 
About channels, we presented it to customer saying: "you can subscribe to stable/beta/alpha channels and get automatic upgrades or... continue to manually manage versions". This looked much more like a step forward than backward in term of maintenance and operations.